presto-native-execution/presto_cpp/main/operators/TestingPersistentShuffle.cpp

```diff
@@ -0,0 +1,145 @@
+ * limitations under the License.
+ */
+#include "presto_cpp/main/operators/TestingPersistentShuffle.h"
+
+using namespace facebook::velox::exec;
+using namespace facebook::velox;
+
+namespace {
+inline std::string createShuffleFilename(
+    const std::string& rootPath,
+    int32_t partition,
+    int fileIndex) {
+  return fmt::format("{}/{}_{}.bin", rootPath, partition, fileIndex);
+}
+
+/// This file is used to indidate that the shuffle system is ready to
+/// be used for reading (acts as a sync point between readers if needed).
+/// Mostly used for test purposes.
+const static std::string kReadyForReadFilename = "readyForRead";
+}; // namespace
+
+namespace facebook::presto::operators {
+
+std::unique_ptr<velox::WriteFile> TestingPersistentShuffle::getNextOutputFile(
+    int32_t partition) {
+  auto fileCount = getPartitionFilesCount(partition);
+  std::string filename = createShuffleFilename(rootPath_, partition, fileCount);
+  return fileSystem_->openFileForWrite(filename);
+}
+
+int TestingPersistentShuffle::getPartitionFilesCount(int32_t partition) const {
+  auto fileCount = 0;
+  bool fileExist = false;
+  do {
+    std::string filename =
+        createShuffleFilename(rootPath_, partition, fileCount);
+    if (!fileSystem_->exists(filename)) {
+      break;
+    }
+    fileCount++;
+  } while (fileExist);
+  return fileCount;
+}
+
+void TestingPersistentShuffle::storePartitionBlock(int32_t partition) {
+  auto& buffer = inProgressPartitions_[partition];
+  buffer->setSize(inProgressSizes_[partition]);
+  auto file = getNextOutputFile(partition);
+  file->append(std::string_view(buffer->as<char>(), buffer->size()));
+  file->close();
+  inProgressPartitions_[partition].reset();
+}
+
+void TestingPersistentShuffle::collect(
+    int32_t partition,
+    std::string_view data) {
+  auto& buffer = inProgressPartitions_[partition];
+
+  // Check if there is enough space in the buffer.
+  if (buffer &&
+      inProgressSizes_[partition] + data.size() + sizeof(size_t) >=
+          buffer->capacity()) {
+    storePartitionBlock(partition);
+  }
+
+  // Allocate buffer if needed.
+  if (!buffer) {
+    buffer = AlignedBuffer::allocate<char>(
+        std::max(uint32_t(data.size()), maxBytesPerPartition_), pool_);
+    inProgressSizes_[partition] = 0;
+  }
+
+  // Copy data.
+  auto rawBuffer = buffer->asMutable<char>();
+  auto offset = inProgressSizes_[partition];
+
+  *(size_t*)(rawBuffer + offset) = data.size();
+
+  offset += sizeof(size_t);
+  memcpy(rawBuffer + offset, data.data(), data.size());
+
+  inProgressSizes_[partition] += sizeof(size_t) + data.size();
+}
+
+void TestingPersistentShuffle::noMoreData(bool success) {
+  // Delete all shuffle files on failure
+  if (!success) {
+    cleanup();
+  }
+  for (auto i = 0; i < numPartitions_; ++i) {
+    if (inProgressSizes_[i] > 0) {
+      storePartitionBlock(i);
+    }
+  }
+  // Writing "ready to read" file
+  fileSystem_->openFileForWrite(
```

Comment:
nit: do we rely on file->close() to be called by destructor? Might be cleaner to add an explicit call. Would it easier to read.

Reply:
Sounds very good